specification/confidentialledger/data-plane/Microsoft.ConfidentialLedger/stable/2022-11-23/confidentialledger.json

```diff
@@ -0,0 +1,1021 @@
+            "description": "Error response describing why the operation failed",
+            "schema": {
+              "$ref": "common.json#/definitions/ConfidentialLedgerError"
+            }
+          }
+        }
+      },
+      "get": {
+        "operationId": "GetUser",
+        "x-ms-examples": {
+          "GetUser": {
+            "$ref": "./examples/ledger/GetUser.json"
+          }
+        },
+        "summary": "Gets a user.",
+        "parameters": [
+          {
+            "$ref": "common.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/UserIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Details about the specified user.",
+            "schema": {
+              "$ref": "#/definitions/LedgerUser"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed",
+            "schema": {
+              "$ref": "common.json#/definitions/ConfidentialLedgerError"
+            }
+          }
+        }
+      },
+      "patch": {
+        "operationId": "CreateOrUpdateUser",
+        "x-ms-examples": {
+          "CreateOrUpdateUser": {
+            "$ref": "./examples/ledger/CreateOrUpdateUser.json"
+          }
+        },
+        "summary": "Adds a user or updates a user's fields.",
+        "description": "A JSON merge patch is applied for existing users",
+        "consumes": [
+          "application/merge-patch+json"
+        ],
+        "parameters": [
+          {
+            "$ref": "common.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/UserIdParameter"
+          },
+          {
+            "$ref": "#/parameters/UserDetailsParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "User created or updated successfully.",
+            "schema": {
+              "$ref": "#/definitions/LedgerUser"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed",
+            "schema": {
+              "$ref": "common.json#/definitions/ConfidentialLedgerError"
+            }
+          }
+        }
+      }
+    }
+  },
+  "definitions": {
+    "ApplicationClaims": {
+      "description": "List of application claim objects.",
+      "type": "array",
+      "items": {
+        "$ref": "#/definitions/ApplicationClaimObject"
+      }
+    },
+    "ApplicationClaimData": {
+      "description": "Data used to build an application claim.",
+      "properties": {
+        "collectionId": {
+          "description": "Identifier for a collection id.",
+          "type": "string"
+        },
+        "contents": {
+          "description": "Contents of a ledger entry.",
+          "type": "string"
+        }
+      },
+      "type": "object"
+    },
+    "ApplicationClaimObject": {
```

Comment:
As also mentioned in another [comment](https://github.com/microsoft/TPAL/pull/214#discussion_r1040070507), I'm not convinced that we should commit to this structure right now, given the lengthy discussion we had to decide how to reveal application claims in a receipt. 

There is the possibility that we decide to change again the application claims structure and converting properties from `required` to `optional`, or viceversa, is considered a breaking change for an Azure API. Therefore, I'd prefer to keep the structure more flexible to make future changes easier to carry out. But please let me know if you have a different opinion and we can discuss about it!

> I also seem to remember that `claimData` may not be a full JSON object if we are including the claim but not revealing it. If that is the case, please express it in the model as well.

Nope, `claimData` is supposed to always be a full JSON object. If we don't reveal the claim, we use a different field (`claimDigest`) to express the unrevealed claim (I'm adding it now to the model as I forgot to add it before 😅)

Reply:
Sorry, I didn't fully think through which fields need to be required, just mentioned it for you to think about. I think we want `version` to be required. Then if we have a version 2 that uses new fields, we can add those additional fields as optional without removing the old fields. But the `version` parameter is super important and makes sense to be required because you cannot do any verification without it, as `version` exactly specifies the fields required in the claim as well as the methodology for creating the claim.

And yeah you're right, `claimData` / `claimDigest` are distinct, but you were definitely missing one so glad you got the intent 😆

Reply:
I understand that the `version` field is important but I don't feel like we are fully established on the current structure, and I fear making some fields required now would put some constraints on how we can evolve the model in the future. However, it's also true that the list structure wrapping `ApplicationClaimObjects` is optional too, so changing that later on should not be a big problem theoretically. Based on this, considering that `claimData` / `claimDigest` may be more prone to change (due to the nature of the data we may want to use as claims), I'd say `version` and `secretKey` can both be required properties as you were suggesting in your initial comment.

Reply:
Addressed in https://github.com/Azure/azure-rest-api-specs/pull/21659/commits/45629addd0a0f6ce4ec567f9eb98eee4adaa7c74